### DISCUSSION TITLE
Problems Writing the Equations for Project 2

### DISCUSSION THREAD

**Discussion by hollycormack21** [IBAHCM/RPiR#488] [Archive] [answered]
Hi 

I've written functions for the general diversity function, the species richness and Simpson's diversity for project 2. 
I'm not getting any error messages but the new functions do not match the results from the original functions. 
p2_species_richness always returns with the 299 
p2_simpson's richness almost always return 0.00334482 except for one.pop where it returns 1
p2_q.diversity usually returns 299 even for  rand.pop and rand.10.pop but I get 46 and 147 for quadrant.pop and quadrant.10
repeatedly, 
Also I need to put print within the species richness function and the diversity function to get a return and I'm not sure why that's happening or if adding print is the correct way to fix t. 
I've repeatedly checked my code and I can't work out how I'm writing the wrong equation. 
Any help would be greatly appreciated. 
<img width="1032" alt="Screenshot 2022-01-24 at 00 09 52" src="https://user-images.githubusercontent.com/94694764/150704261-5ea4ff13-7714-407a-86c8-daeb2b523a9f.png">
<img width="1023" alt="Screenshot 2022-01-24 at 00 10 04" src="https://user-images.githubusercontent.com/94694764/150704264-e68a3320-6912-4295-8b45-8e8b97e6857d.png">
<img width="1000" alt="Screenshot 2022-01-24 at 00 10 12" src="https://user-images.githubusercontent.com/94694764/150704269-fa9c616b-f19a-449e-b349-a2b573655a80.png">

**Comment by richardreeve** [comment, marked as answer]
Hi @hollycormack21 - it looks like you are not calculating `prop` correctly in your general function. It you just look at the different populations and see what `prop` is for each of them (i.e. calculate `tree.counts` and then `prop`) then you should see that those are not the population proportions of the non-zero species. If you can get that right, the rest should work better.

**Reply by hollycormack21** [reply]
Hi 
Thank you for your answer. I'm afraid I don't really understand. I thought you were saying there's something wrong with the way I'm calculating prop and so I've tried running the calculations for prop using 
tree.counts <- quadrant.pop <0 
prop <- (tree.counts/ (sum(tree.counts))
However prop returns a list of all the species and their proportions and because some of the populations have 299 species so I'm having a hard time comparing them. 
Is the problem that I'm calculating tree.counts wrong or prop wrong?

**Reply by richardreeve** [reply]
`tree.counts` seems to be a vector telling you which populations are non zero (`FALSE` (=0) if they are absent, `TRUE` (= 1) if they are present). That is probably not what you intended given the name you have given the variable. If you try those two lines of code with a population like `new.pop <- c(2, 1, 0)` and think about what you would expect the population proportions of the species present to be (hint: 2/3 and 1/3, with the third one absent), then you'll see you don't get this. This is the kind of thing you need to be thinking about - but I can't tell you how to do all of the exercises, you need to work out what the correct approach is yourself.
